pkg/stanza/fileconsumer/file.go

```diff
@@ -213,17 +222,39 @@ func (m *Manager) makeReaders(ctx context.Context, paths []string) {
 			continue
 		}
 
-		r, err := m.newReader(ctx, file, fp)
+		r, matchFound, err := m.newReader(ctx, file, fp)
 		if err != nil {
 			m.set.Logger.Error("Failed to create reader", zap.Error(err))
 			continue
 		}
+		if matchFound {
+			m.tracker.Add(r)
+		} else {
+			m.unmatchedFiles = append(m.unmatchedFiles, archive.NewArchiveRecord(file, fp))
+		}
+	}
 
-		m.tracker.Add(r)
+	readers, err := m.archive.Match(m.unmatchedFiles)
+	if err != nil {
+		m.set.Logger.Error("Errors encountered while creating readers", zap.Error(err))
 	}
+	for _, reader := range readers {
+		// Exclude duplicate paths with the same content. This can happen when files are
+		// being rotated with copy/truncate strategy. (After copy, prior to truncate.)
+		if r := m.tracker.GetCurrentFile(reader.Fingerprint); r != nil {
+			m.set.Logger.Debug("Skipping duplicate file", zap.String("path", r.GetFileName()))
+			// re-add the reader as Match() removes duplicates
+			m.tracker.Add(r)
+			reader.Close()
+			continue
+		}
+		m.tracker.Add(reader)
+		m.set.Logger.Info("Started watching file", zap.String("path", reader.GetFileName()))
+	}
+
```

Comment:
I don't like the way this moves some of the matching & reader creation logic into a different place. Can it just stay within `newReader`?